src/index.d.ts

```diff
@@ -0,0 +1,55 @@
+import React from 'react';
+
+type BoxOptions = 'border-box' | 'content-box' | 'device-pixel-content-box';
+
+interface ResizeObserverBoxSize {
+  readonly inlineSize: number;
+  readonly blockSize: number;
+}
+
+interface ResizeObserverEntry {
+  readonly target: Element;
+  readonly contentRect: DOMRectReadOnly;
+  readonly borderBoxSize: Array<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+  readonly contentBoxSize: Arrray<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+  readonly devicePixelContentBoxSize: Arrray<ResizeObserverBoxSize> | ResizeObserverBoxSize;
+}
+
+interface Breakpoints<T> {
+  [key: number]: T;
+}
+
+interface BreakpointsOptions<W, H> {
+  box?: BoxOptions;
+  widths?: Breakpoints<W>;
+  heights?: Breakpoints<H>;
+  fragment?: number;
+}
+
+interface ObservedElementProps<T> {
+  ref: React.RefObject<T>;
+}
+
+interface ObserveRenderArgs<E, W, H> {
+  observedElementProps: ObservedElementProps<E>;
+  widthMatch: W;
+  heightMatch: H;
+}
+
+interface ObserveProps<E, W, H> {
+  box?: BoxOptions;
+  breakpoints?: BreakpointsOptions<W, H>;
+  render: ({
+    observedElementProps,
+    widthMatch,
+    heightMatch
+  }: ObserveRenderArgs<E, W, H>) => JSX.Element;
+}
+
+export const Observe: <E, W, H>(props: ObserveProps<E, W, H>) => JSX.Element;
```

Comment:
This component accepts the following props:
```javascript
<Observe
  box='border-box' // or `undefined`, 'content-box', 'device-pixel-content-box'
  breakpoints={/* same interface as useBreakpoints */}
  render={renderFunction}
/>
```
`box` here is _not_ the same as the one used in the `options` object for the `breakpoints` prop (or `useBreakpoints()`, for that matter). It instructs the ResizeObserver which box of the element to watch. It has the same signature, though, so it can be undefined or any of the [draft spec](https://drafts.csswg.org/resize-observer/#api) box sizes.

`breakpoints` accepts an `options` object identical to `useBreakpoints()`. In fact, it's passed on to `useBreakpoints()` in this component internally.

`render` accepts a function which will be used to render your DOM children. It wraps them in a `React.Context` and calls the function that you pass in. You'll get access to the following arguments:
```javascript
const renderFunction = ({
  observedElementProps,
  widthMatch,
  heightMatch
}) => { ... }
```
`observedElementProps` is an object with the shape `{ ref: React.RefObject }`. You use it to apply `ref` to a DOM element, and possibly other props in the future, like `<div {...observedElementProps}>`.

`widthMatch` and `heightMatch` are only useful when you are using the `breakpoints` prop. If you don't use that prop, these will always be `undefined`. Otherwise, they will contain the values that `useBreakpoints()` returns: `[widthMatch, heightMatch]`.